File: internalUsages/SharedDomain/Aggregates/User/UserProjector.cs
using Sekiban.Pure.Aggregates;
using Sekiban.Pure.Events;
using Sekiban.Pure.Projectors;
using SharedDomain.Aggregates.User.Events;

namespace SharedDomain.Aggregates.User;

public class UserProjector : IAggregateProjector
{
    public IAggregatePayload Project(IAggregatePayload payload, IEvent ev)
    {
        var user = payload as User ?? new User(Guid.Empty, string.Empty, string.Empty);
        
        return ev switch
        {
            Event<UserCreated> e => new User(e.Payload.UserId, e.Payload.Name, e.Payload.Email),
            Event<UserNameChanged> e => user with { Name = e.Payload.NewName },
            Event<UserEmailChanged> e => user with { Email = e.Payload.NewEmail },
            _ => user
        };
    }
}

public record User(
    Guid UserId,
    string Name,
    string Email) : IAggregatePayload;

File: ts/samples/dapr-sample/packages/domain/src/aggregates/user/projectors/user-projector.ts
import { defineProjector, EmptyAggregatePayload, type ITypedAggregatePayload } from '@sekiban/core';
import { z } from 'zod';
import { UserCreated, UserNameChanged, UserEmailChanged } from '../events/index.js';

// User state schema
const userStateSchema = z.object({
  aggregateType: z.literal('User'),
  userId: z.string().uuid(),
  name: z.string(),
  email: z.string().email(),
  createdAt: z.string(),
  updatedAt: z.string()
});

export interface UserState extends ITypedAggregatePayload {
  readonly aggregateType: 'User';
  userId: string;
  name: string;
  email: string;
  createdAt: string;
  updatedAt: string;
}

// Define the User projector
export const UserProjector = defineProjector({
  aggregateType: 'User',
  initialState: () => new EmptyAggregatePayload(),
  projections: {
    [UserCreated.type]: (state, event) => {
      const userState: UserState = {
        aggregateType: 'User',
        userId: event.payload.userId,
        name: event.payload.name,
        email: event.payload.email,
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString()
      };
      return userState;
    },
    [UserNameChanged.type]: (state, event) => {
      return {
        ...state,
        name: event.payload.newName,
        updatedAt: new Date().toISOString()
      };
    },
    [UserEmailChanged.type]: (state, event) => {
      return {
        ...state,
        email: event.payload.newEmail,
        updatedAt: new Date().toISOString()
      };
    }
  }
});